CIPs/0017.md

```diff
@@ -0,0 +1,165 @@
+that node A currently *has* and requests those it is missing.
+Node B's reply must contain any proofs that node B has that were not included in the message node A sent.
+If `complement` is false, then this message is sent with the set of proofs that node A currently
+*does not have* and requests these from this peer.
+Node B's reply must contain the proofs requested, if node B has them. Node A should only request proofs
+that node B has indicated via `NewPlumoProof` that it has.
+
+
+PlumoProofMsg (0x1b)
+[plumoProof_0, plumoProof_1, ...]
+Reply to GetPlumoProofMsg. The items in the list are plumo proof structs, {proof: []byte, firstEpoch: uint, lastEpoch: uint, versionNumber: uint}.
+This may validly contain no plumo proofs if the GetPlumoProofMsg included all proofs this node is aware of.
+```
+
+This will be accompanied by a new `proofFetcher` object to be in charge of fetching these proofs from peers while providing protections against DoS attacks. This feels like overkill, but seems necessary to prevent attacks and maintain state.
+
+### `les` protocol changes
+
+Relevant proposed structs:
+```
+// Same as in the `eth` protocol, minimal proof metadata
+ProofMetadata struct {
+  FirstEpoch uint
+  LastEpoch uint
+  VersionNumber uint
+}
+
+// Metadata for each epoch to construct `EpochBlock` for the snark's verify fn
+// May be unnecessary if we can make assumptions on `MaxNonSigners`
+LightEpochBlock struct {
+  Index         uint // Epoch Index 
+  MaxNonSigners uint // MinQuorumSize()
+}
+
+// All of the data needed by the light client to verify and utilize a proof.
+// This is also crafted to allow for batch verifications using `FirstHashToField`
+// My rough calculations show this struct is ~439 bytes for a non-batched proof with no val set differences.
+LightPlumoProof struct {
+  Proof             []byte // Serialized proof. Latest estimate from @psi = 383 bytes
+  FirstEpoch        LightEpochBlock // First epoch this proof is over
+  LastEpoch         LightEpochBlock // Last epoch this proof is over
+  VersionNumber     uint // Proof version number, indicates the corresponding verification key to use
+  LastHashToField  []byte  // Hash to field of Last epoch's `snark.EpochBlock` struct. To be used for batch verification.
+                           // If there is no batch verification, this will be empty. May need an additional boolean to signify this.
+                           // If batch verification should be used, newValidators and deletedValidators will be empty.
+  NewValidators     []blsCrypto.SerializedPublicKey // ValSetDiff(FirstEpoch, LastEpoch), added validators
+  DeletedValidators *big.Int // ValSetDiff(FirstEpoch, LastEpoch): bitmap of deleted validators from the `firstEpoch` valset.
+}
+```
+
+The light clients will need to store verification keys corresponding to the existing versions of the snark circuit, as elaborated in #upgradeability. 
+
+There are two proposed extensions for the `les` protocol:
+
+Option 1:
```

Comment:
After coming back to this I'm having trouble understanding why this handles incomplete proof propagation better than approach 2? The only case I've thought of is where this approach is better is when the full node would have proofs that start more than 240 epochs ago, in which case we'd presumably just ask a different full node.

Reply:
I spoke with @nategraf about this yesterday, one specific case he brought up that I hadn't thought of is the "release process" for new circuits.
Say we upgrade the circuit to v2, this implies a new VK that full nodes must receive in order to validate and accept/propagate further proofs. Depending on how this is implemented, we could be left in a situation where some full nodes have more proofs by nature of having v2 proofs.
The end goal is of course is to allow light clients to verify they are syncing via the fastest available method, so this "proofInventory" is one way, another would be to receive all available proofs from all peers and then follow the shortest path, but that seems even more cumbersome to implement.